e2e/jest.test-setup.ts

```diff
@@ -8,79 +10,170 @@ const userAgent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/5
+      // truncate long response. get first two workspace details.
+      responseText = fp.isEmpty(JSON.parse(responseText).items)
+         ? responseText
+         : 'truncated...\n' + JSON.stringify(JSON.parse(responseText).items.slice(0, 2), null, 2);
+    }
+    console.debug('❗Request finished: ' +
+       `${request.response().status()} ${request.method()} ${request.url()}\n${responseText}`);
+  }
+
+  const isWorkbenchRequest = fp.flow(isWorkbenchApi, notOptionsRequest, includeResourceType, includeUrl);
+  const getRequestData = fp.flow(getRequestPostData, stringifyData);
+  const canLogResponse = fp.flow(isWorkbenchRequest, notRedirectRequest);
+
+  // New request initiated
   page.on('request', (request) => {
+    if (isWorkbenchRequest(request)) {
+      console.debug('❗Request issued: ' +
+         `${request.method()} ${request.url()}\n${getRequestData(request)}`);
+    }
     try {
-      if (request.url().includes('api-dot-all-of-us')) {
-        const method = request.method();
-        if (method !== 'OPTIONS') {
-          console.debug(`❗Request issued: ${request.url()}`);
-        }
-      }
       request.continue();
       // tslint:disable-next-line:no-empty
     } catch (e) {
     }
-  })
-  .on('requestfinished', async (request) => {
-    try {
-      const response = await request.response();
-      const status = request.response().status();
-      let responseBody;
-      if (request.redirectChain().length === 0) {
-        // response body can only be accessed for non-redirect responses.
-        if (request.url().includes('api-dot-all-of-us')){
-          responseBody = await response.buffer();
-          if (responseBody !== undefined) {
-            responseBody = JSON.stringify(JSON.parse(responseBody.toString()), null, 2);
-          }
-          const method = request.method();
-          console.debug(`❗Request finished: ${status} ${method} ${request.url()}  \n ${responseBody}`, {depth: null, colors: true});
-        }
+  });
+
+  page.on('requestfinished', async (request) => {
+    if (canLogResponse(request)) {
+      if (isApiFailure(request)) {
+        await logError(request);
+      } else {
+        await logResponse(request);
       }
+    }
+    try {
       await request.continue();
       // tslint:disable-next-line:no-empty
     } catch (e) {
     }
-  })
-  .on('requestfailed', async (request) => {
+  });
+
+  page.on('console', async (message) => {
+    if (!message.args().length) {
+      return;
+    }
+    const title = await getTitle();
     try {
-      const response = request.response();
-      if (response !== null) {
-        const status = response.status();
-        const responseText = await response.text();
-        const failureError = request.failure().errorText;
-        console.debug(`❗Request failed: ${status} ${request.method()} ${request.url()}  \n ${failureError} \n ${responseText}`);
-      }
+      const texts = await Promise.all(
+        message.args().map(async (arg) =>
```

Comment:
nit: To keep the code consistent, it may be best to use `fp.map` instead - it will prevent mental context switching between native and fp functions in the future

Reply:
I tried but failed to make `fp.map` to work with async functions inside of.